test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp

```diff
@@ -21,210 +21,219 @@
 #include "support/test_config.h"
 #include "support/utils.h"
 
-#include <iostream>
-#include <iomanip>
-
-template<typename _T1, typename _T2>
-void ASSERT_EQUAL(_T1&& X, _T2&& Y) {
-    if(X!=Y)
-        std::cout << "CHECK CORRECTNESS (REDUCE_BY_SEGMENT): fail (" << X << "," << Y << ")" << std::endl;
-}
-
 #if TEST_DPCPP_BACKEND_PRESENT
-#include <CL/sycl.hpp>
+#    include <CL/sycl.hpp>
+
+using namespace oneapi::dpl::execution;
+#endif
+using namespace TestUtils;
 
-void test_with_buffers()
+struct test_reduce_by_segment
 {
-    // create buffers
-    sycl::buffer<uint64_t, 1> key_buf{ sycl::range<1>(13) };
-    sycl::buffer<uint64_t, 1> val_buf{ sycl::range<1>(13) };
-    sycl::buffer<uint64_t, 1> key_res_buf{ sycl::range<1>(13) };
-    sycl::buffer<uint64_t, 1> val_res_buf{ sycl::range<1>(13) };
 
+    template <typename Iterator1, typename Iterator2, typename Iterator3, typename Iterator4, typename Size>
+    void
+    initialize_data(Iterator1 host_keys, Iterator2 host_vals, Iterator3 host_key_res, Iterator4 host_val_res, Size n)
     {
-        auto keys    = key_buf.template get_access<sycl::access::mode::read_write>();
-        auto vals    = val_buf.template get_access<sycl::access::mode::read_write>();
-        auto keys_res    = key_res_buf.template get_access<sycl::access::mode::read_write>();
-        auto vals_res    = val_res_buf.template get_access<sycl::access::mode::read_write>();
-
-        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 };
-        //T vals[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 };
-
-        // keys_result = {1, 2, 3, 4, 1, 3, 1, 3, 0};
-        // vals_result = {1, 2, 3, 4, 2, 6, 2, 6, 0};
-
-        // Initialize data
-        for (int i = 0; i != 12; ++i) {
-            keys[i] = i % 4 + 1;
-            vals[i] = i % 4 + 1;
-            keys_res[i] = 9;
-            vals_res[i] = 1;
-            if (i > 3) {
-                ++i;
-                keys[i] = keys[i-1];
-                vals[i] = vals[i-1];
-                keys_res[i] = 9;
-                vals_res[i] = 1;
+        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, ..., 0 };
+        //T vals[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, ..., 0 };
+        for (int i = 0; i<n - 1; i> 3 ? i += 2 : ++i, ++host_keys, ++host_vals, ++host_key_res, ++host_val_res)
+        {
+            *host_keys = i % 4 + 1;
+            *host_vals = i % 4 + 1;
+            *host_key_res = 9;
+            *host_val_res = 1;
+            if ((i > 3) && (i + 1 < n - 1))
+            {
+                auto tmp = *host_keys;
+                ++host_keys;
+                *host_keys = tmp;
+                tmp = *host_vals;
+                ++host_vals;
+                *host_vals = tmp;
+                ++host_key_res;
+                ++host_val_res;
+                *host_key_res = 9;
+                *host_val_res = 1;
             }
         }
-        keys[12] = 0;
-        vals[12] = 0;
+        *host_keys = 0;
+        *host_vals = 0;
```

Comment:
I think it would be much readable:
```c++
    for(std::size_t i = 0; i < n - 1; ++i)
    {
        std::size_t fill_first = i + 1;
        std::size_t fill_repeat = i % 4 < 2 ? 1 : 3;
        std::size_t fill = i < 4 ? fill_first : fill_repeat;

        *host_vals++ = fill;
        *host_keys++ = fill;
        *host_key_res++ = 9;
        *host_val_res++ = 1;
    }
    *host_vals = 0;
    *host_keys = 0;
    *host_key_res = 0;
    *host_val_res = 0;
```

Reply:
Agreed.  Thanks for the simplification. Done.